Remove Deprecated Code

JonathonMisiewicz opened the pull request:
## Description
This PR removes some deprecated code, especially regarding molden.


## User API & Changelog headlines
<!-- A bullet-point format description of how this PR affects the user.
     This is destined for the release notes. May be empty. -->
- [x] Fixed a bug that would cause a crash when calling `wfn.write_molden` on a non-HF wavefunction without marking natural orbitals.
- [x] The deprcated `dcft` keyword is no longer recognized.  
- [x] `dfocc`'s `molden_write` is now deprecated. Just call the `molden_write` method of the associated wavefunction.
- [x] `Matrix::Full` is removed in favor of `Matrix::SubBlocks`, as the deprecation notice foretold. 
- [x] Changes to CheMPS2 to allow working with new Molden code.
- [x] Behavior Change: In `psi4.molden`, if densities are not provided, if `occupation_a` and `occupation_b` are not found, these will be pulled from `reference_wavefunction` instead of set to zero.
- [x] Behavior Change: `wfn.write_molden` will only check that the orbital energies agree if the molden writing actually calls for orbital energies.

## Dev notes & details
<!-- A bullet-point format description of what this PR does "at a glance."
     Target audience is code reviewers and other devs skimming PRs.
     Should be more technical than user notes. Should never be empty. -->
- [x] `psi4.molden` is now a wrapper on `wfn.write_molden`.
- [x] Migrated `dfocc` and `dmrg` from using C-Molden writer to the Python wavefunction method.
- [x] C-Molden writer (buggy, deprecated several versions ago) is deleted.
- [x] DMRGSolver is now a wavefunction.
- [x] DMRGSolver is equipped with occupation_a and occupation_b that store the diagonal of the MO RDM in the current orbitals, mimicking SCF.
- [x] DMRGSolver's `build_rdm_ao` has been re-organized
- [x] DMRG with pseudocanonical orbitals now updates `epsilon_a_` and `epsilon_b_` accordingly.
- [x] The MO basis MS-AVERAGED DMRG density is added to the density map.
- [x] Compiler warnings in `libmints` are silenced.
- [x] Some deprecated miscellanea are removed. 


## Questions
- [ ] Tests wanted?

## Checklist
- [x] molden and CheMPS2 tests pass

## Status
- [x] Ready for review
- [x] Ready for merge

JonathonMisiewicz commented:
I'll deal with this in the morning...
Note to self: there appear to be two issues. One is with GauXC (not sure what to do about that one), but the other is symbols not being found... Probably, the missing symbols are due to missing `ifdef USING_CheMPS2` - we ought never to go searching for those symbols.

loriab commented on psi4/src/core.cc:
```diff
@@ -572,6 +560,9 @@ void throw_deprecation_errors(std::string const& key, std::string const& module
         py_psi_print_out("WARNING!\n\tRemove keyword PK_ALGO! PK_ALGO has been replaced by the SCF_SUBTYPE=YOSHIMINE_OUT_OF_CORE and REORDER_OUT_OF_CORE options. Using PK_ALGO will raise an error in v1.8.\n");
     }
+    if (module == "DFOCC" && key == "MOLDEN_WRITE") {
+        py_psi_print_out("WARNING!\n\tRemove keyword MOLDEN_WRITE! Using MOLDEN_WRITE will raise an error in v1.12. Call wrn.write_molden instead.\n");
```
```suggestion
        py_psi_print_out("WARNING!\n\tRemove keyword MOLDEN_WRITE! Using MOLDEN_WRITE will raise an error in v1.12. Call wfn.write_molden instead.\n");
```

loriab reviewed (commented):
I need to read it through again, but overall lgtm. lmk if you need any advice with the `#ifdef USING_CheMPS2` etc. or if CI turns up anything odd.

JonathonMisiewicz commented on psi4/src/psi4/libsapt_solver/amplitudes.cc:
```diff
@@ -1158,21 +1159,19 @@ void SAPT2p3::ind30_amps(int AAfile, const char *ARlabel, int BBfile, const char
     for (int a = 0; a < noccA; a++) {
         for (int r = 0; r < nvirA; r++) {
-            uAR[a][r] /= evalsA[a] - evalsA[r + noccA];
+            uAR.get_pointer()[r] /= evalsA[a] - evalsA[r + noccA];
```
I would expect it _would_. I'll figure out why tests didn't catch this and add a new test if needed.

loriab replied:
hooray, best case. glad sapt has good coverage.

JonathonMisiewicz commented:
> I need to read it through again, but overall lgtm. lmk if you need any advice with the `#ifdef USING_CheMPS2` etc. or if CI turns up anything odd.

The ifdef cures the build problems, locally. It remains to be seen if CI turns up anything else odd... I will let you know when I've pushed my local changes. I have one last thing to investigate.

loriab commented on psi4/driver/driver.py:
```diff
@@ -2050,51 +2049,27 @@ def molden(wfn, filename=None, density_a=None, density_b=None, dovirtual=None):
-        mw = core.MoldenWriter(wfn)
-        mw.write(filename, NO_Ca, NO_Cb, NO_occa, NO_occb, NO_occa, NO_occb, dovirt)
+            wfn_.Db().copy(wfn_.Da())
+        wfn_.write_molden(filename, dovirtual, True)
```
(comment has no relation to line of code)

there's a couple places in the docs to amend (search "molden"):
* https://github.com/psi4/psi4/blob/master/doc/sphinxman/source/psithoninput.rst
* doc/sphinxman/source/molden.rst (you can try autofunction psi4.core.Wavefunction.write_molden

loriab reviewed (commented):
Have to go to a mtg, but I've read most.

JonathonMisiewicz commented on psi4/src/psi4/libsapt_solver/amplitudes.cc:
```diff
@@ -1158,21 +1159,19 @@ void SAPT2p3::ind30_amps(int AAfile, const char *ARlabel, int BBfile, const char
-    psio_->write_entry(ampout, amplabel, (char *)uAR[0], sizeof(double) * noccA * nvirA);
-
-    free_block(uAR);
+    uAR.save(psio_, ampout, Matrix::SaveType::SubBlocks); 
```
Correct. The purpose of `reset` is to free the memory. I'm happy to let smart pointers be smart.

loriab reviewed (approved):
Thanks for the great cleanup and consolidation

JonathonMisiewicz commented:
After talking with Lori, we decided _not_ to deprecate `psi4.molden`. I removed a wrong deprecation warning and updated the PR overview accordingly. Sorry for the confusion.
